Library/DiscUtils.Ext/ExtentsFileBuffer.cs

```diff
@@ -70,36 +70,46 @@ public override int Read(long pos, byte[] buffer, int offset, int count)
+                uint logicalBlock = (uint)((pos + totalRead) / blockSize);  // logical block containing the next byte to read
+                int blockOffset = (int)(pos + totalRead - logicalBlock * blockSize);  // offset within 'logicalBlock' of the next byte to read
 
                 int numRead = 0;
 
+                // Find the extent containing 'logicalBlock' or, if none, the first extent beyond it.
                 Extent extent = FindExtent(extents, logicalBlock);
+
                 if (extent == null)
                 {
-                    throw new IOException("Unable to find extent for block " + logicalBlock);
+                    // The remainder of the sparse file is one big hole; all bytes remaining to be read in the file are zeroes.
+                    numRead = totalBytesRemaining;
+                    Array.Clear(buffer, offset + totalRead, numRead);
                 }
-                if (extent.FirstLogicalBlock > logicalBlock)
+                else if (extent.FirstLogicalBlock > logicalBlock)
                 {
-                    numRead =
-                        (int)
+                    // We're reading from a hole in the sparse file that ends at 'extent'.  Implicitly read zeroes
+                    // for the remainder of 'logicalBlock', and for all subsequent blocks before the beginning of 'extent'.
+                    numRead = (int)
                         Math.Min(totalBytesRemaining,
-                            (extent.FirstLogicalBlock - logicalBlock) * blockSize - blockOffset);
+                                 (extent.FirstLogicalBlock - logicalBlock) * blockSize - blockOffset);
                     Array.Clear(buffer, offset + totalRead, numRead);
                 }
                 else
                 {
-                    long physicalBlock = logicalBlock - extent.FirstLogicalBlock + (long)extent.FirstPhysicalBlock;
-                    int toRead =
-                        (int)
+                    // We found the extent containing 'logicalblock'.  Read all bytes from 'logicalBlock' until the end of this extent.
+                    int toRead = (int)
                         Math.Min(totalBytesRemaining,
-                            (extent.NumBlocks - (logicalBlock - extent.FirstLogicalBlock)) * blockSize - blockOffset);
+                                 (extent.FirstLogicalBlock + extent.NumBlocks - logicalBlock) * blockSize - blockOffset);
 
+                    long physicalBlock = logicalBlock - extent.FirstLogicalBlock + (long)extent.FirstPhysicalBlock;
                     _context.RawStream.Position = physicalBlock * blockSize + blockOffset;
                     numRead = _context.RawStream.Read(buffer, offset + totalRead, toRead);
                 }
 
+                if (numRead == 0)
+                {
+                    throw new IOException($"Unable to read logical block {logicalBlock};  extent?.FirstLogicalBlock: {extent?.FirstLogicalBlock}");
```

Comment:
This doesn't fix reading correctness, but will avoid hang, right?

Reply:
Added a comment to explain intent.